File: chrome/browser/resources/settings/prefs/pref_util.js
// Copyright 2015 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

/** @fileoverview Utility functions to help use prefs in Polymer controls. */

// TODO(michaelpg): converge with other WebUI on capitalization. This is
// consistent with Settings, but WebUI uses lower.underscore_case.
cr.define('Settings.PrefUtil', function() {
  /**
   * Converts a string value to a type corresponding to the given preference.
   * @param {string} value
   * @param {!chrome.settingsPrivate.PrefObject} pref
   * @return {boolean|number|string|undefined}
   */
  function stringToPrefValue(value, pref) {
    switch (pref.type) {
      case chrome.settingsPrivate.PrefType.BOOLEAN:
        return value == 'true';
      case chrome.settingsPrivate.PrefType.NUMBER:
        var n = parseInt(value, 10);
        if (isNaN(n)) {
          console.error(
              'Argument to stringToPrefValue for number pref ' +
              'was unparsable: ' + value);
          return undefined;
        }
        return n;
      case chrome.settingsPrivate.PrefType.STRING:
      case chrome.settingsPrivate.PrefType.URL:
        return value;
      default:
        assertNotReached('No conversion from string to ' + pref.type + ' pref');
    }
  }

  /**
   * Returns the value of the pref as a string.
   * @param {!chrome.settingsPrivate.PrefObject} pref
   * @return {string}
   */
  function prefToString(pref) {
    switch (pref.type) {
      case chrome.settingsPrivate.PrefType.BOOLEAN:
      case chrome.settingsPrivate.PrefType.NUMBER:
        return pref.value.toString();
      case chrome.settingsPrivate.PrefType.STRING:
      case chrome.settingsPrivate.PrefType.URL:
        return /** @type {string} */ (pref.value);
      default:
        assertNotReached('No conversion from ' + pref.type + ' pref to string');
    }
  }
  return {
    stringToPrefValue: stringToPrefValue,
    prefToString: prefToString,
  };
});

File: src/components/ntp_snippets/pref_util.cc
// Copyright 2016 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "components/ntp_snippets/pref_util.h"

#include <memory>

#include "base/values.h"
#include "components/prefs/pref_service.h"

namespace ntp_snippets {
namespace prefs {

std::set<std::string> ReadDismissedIDsFromPrefs(const PrefService& pref_service,
                                                const std::string& pref_name) {
  std::set<std::string> dismissed_ids;
  const base::Value* list = pref_service.GetList(pref_name);
  for (const base::Value& value : list->GetList()) {
    DCHECK(value.is_string())
        << "Failed to parse dismissed id from prefs param " << pref_name
        << " into string.";
    dismissed_ids.insert(value.GetString());
  }
  return dismissed_ids;
}

void StoreDismissedIDsToPrefs(PrefService* pref_service,
                              const std::string& pref_name,
                              const std::set<std::string>& dismissed_ids) {
  base::ListValue list;
  for (const std::string& dismissed_id : dismissed_ids) {
    list.Append(dismissed_id);
  }
  pref_service->Set(pref_name, list);
}

}  // namespace prefs
}  // namespace ntp_snippets
